csi_camera.cpp

```diff
@@ -0,0 +1,367 @@
+#include <iostream>
+#include <fstream>
+#include <cstring>
+#include <sstream>
+#include <chrono>
+#include <thread>
+
+#include "constraints.h"
+#include "csi_camera.h"
+
+using namespace viam::sdk;
+
+CSICamera::CSICamera(const std::string name, const AttributeMap attrs) : Camera(std::move(name)) {
+    std::cout << "Creating CSICamera with name: " << name << std::endl;
+    init(attrs);
+}
+
+CSICamera::~CSICamera() {
+    std::cout << "Destroying CSICamera" << std::endl;
+    stop_pipeline();
+}
+
+void CSICamera::init(const AttributeMap attrs) {
+    validate_attrs(attrs);
+    auto pipeline_args = create_pipeline();
+    if (debug) {
+        std::cout << "pipeline_args: " << pipeline_args << std::endl;
+    }
+    init_csi(pipeline_args);
+}
+
+void CSICamera::validate_attrs(const AttributeMap attrs) {
+    set_attr<int>(attrs, "width_px", &CSICamera::width_px, DEFAULT_INPUT_WIDTH);
+    set_attr<int>(attrs, "height_px", &CSICamera::height_px, DEFAULT_INPUT_HEIGHT);
+    set_attr<int>(attrs, "frame_rate", &CSICamera::frame_rate, DEFAULT_INPUT_FRAMERATE);
+    set_attr<std::string>(attrs, "video_path", &CSICamera::video_path, DEFAULT_INPUT_SENSOR);
+    set_attr<bool>(attrs, "debug", &CSICamera::debug, false);
+}
+
+template <typename T>
+void CSICamera::set_attr(const AttributeMap& attrs, const std::string& name, T CSICamera::* member, T de) {
+    if (attrs->count(name) == 1) {
+        std::shared_ptr<ProtoType> proto = attrs->at(name);
+        auto val = proto->proto_value();
+
+        if constexpr (std::is_same<T, int>::value) {
+            this->*member = val.has_number_value() ? val.number_value() : de;
+        } else if constexpr (std::is_same<T, std::string>::value) {
+            this->*member = val.has_string_value() ? val.string_value() : de;
+        } else if constexpr (std::is_same<T, bool>::value) {
+            this->*member = val.has_bool_value() ? val.bool_value() : de;
+        }
+    } else {
+        this->*member = de; // Set the default value if the attribute is not found
+    }
+}
+
+void CSICamera::reconfigure(const Dependencies deps, const ResourceConfig cfg) {
+    if (debug) {
+        std::cout << "Reconfiguring CSI Camera module" << std::endl;
+    }
+    stop_pipeline();
+    auto attrs = cfg.attributes();
+    init(attrs);
+}
+
+Camera::raw_image CSICamera::get_image(const std::string mime_type) {
+    if (debug) {
+        std::cout << "hit get_image. expecting mime_type " << mime_type << std::endl;
+    }
+    raw_image image;
+    image.mime_type = DEFAULT_OUTPUT_MIMETYPE;
```

Comment:
Why are we not setting this to the given `mime_type`? It looks like that argument is ignored. If so, the debug log is misleading since we aren't expecting the MIME type given in the argument but always the default one.

Reply:
Currently do not have a way to change mime type based on server request. So we always set the mimetype to default (jpeg).